### PR TITLE
Always require closure parameters to be `Sized`

### PR DESCRIPTION
**Pull request by Jules-Bertholet**
The `rust-call` ABI isn't compatible with `#![feature(unsized_fn_params)]`, so trying to use that feature with closures leads to an ICE (#67981). This turns that ICE into a type-check error.

@rustbot label A-closures F-unsized_fn_params

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/closure.rs [hunk 1]
```diff
@@ -89,6 +89,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             expr_def_id,
             body,
             closure.movability,
+            false,
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/closure.rs:92]: comment here please, or ideally a newtyped boolean

File: compiler/rustc_trait_selection/src/traits/select/confirmation.rs [hunk 2]
```diff
@@ -646,18 +646,20 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
...
-            self,
-            obligation.param_env,
-            cause.clone(),
-            obligation.recursion_depth,
-            output_ty,
-            &mut nested,
-        );
-        let tr = ty::TraitRef::from_lang_item(self.tcx(), LangItem::Sized, cause.span, [output_ty]);
-        nested.push(Obligation::new(self.infcx.tcx, cause, obligation.param_env, tr));
+        // Confirm the `Sized` bounds that are present on the argument
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/confirmation.rs:649]: Do we need to do this during confirmation? I think we could probably just register `Sized` obligations when checking a closure during HIR typeck.

**Comment by compiler-errors** [comment]
Can you test more combinations, like `Fn<([u8],)>` obligations with the (desugared) fn trait?

File: compiler/rustc_trait_selection/src/traits/select/confirmation.rs (hunk 2 above)
Comment (Jules-Bertholet) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:649]: Now that I think more about this, I'm not sure there is a way to actually hit this check—the `Sized` bound on the type parameter is checked anyway, to ensure the trait is well-formed. I think I will just remove it

**Comment by compiler-errors** [comment]
Squash this into one PR, please, then I can approve.

**Comment by Jules-Bertholet** [comment]
Done

File: compiler/rustc_trait_selection/src/traits/select/confirmation.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs]: Wait, this definitely still ICEs.

```
#![feature(unsized_fn_params)]

fn bad() -> fn([u8]) { todo!() }

fn main() {
    let f: fn([u8]) = |_| {};
    let slice: Box<[u8]> = Box::new([1; 8]);
    f(*slice);
}
```

Comment (Jules-Bertholet) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs]: > Wait, this definitely still ICEs.

Not on my machine...

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs]: ... I didn't copy and paste that example correctly

**Pull request merged**